core/editor/src/dispatcher/mod.rs

```diff
@@ -9,83 +9,74 @@ pub struct Dispatcher {
+				tool_state.primary_color = *color;
 			}
 			Event::SelectSecondaryColor(color) => {
-				state.tool_state.secondary_color = color;
-
-				Ok(())
+				tool_state.secondary_color = *color;
 			}
 			Event::SwapColors => {
-				std::mem::swap(&mut state.tool_state.primary_color, &mut state.tool_state.secondary_color);
-
-				Ok(())
+				std::mem::swap(&mut tool_state.primary_color, &mut tool_state.secondary_color);
 			}
 			Event::ResetColors => {
-				state.tool_state.primary_color = Color::BLACK;
-				state.tool_state.secondary_color = Color::WHITE;
-
-				Ok(())
+				tool_state.primary_color = Color::BLACK;
+				tool_state.secondary_color = Color::WHITE;
 			}
 			Event::MouseDown(mouse_state) => {
-				state.tool_state.mouse_state = mouse_state;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state = *mouse_state;
 			}
 			Event::MouseUp(mouse_state) => {
-				state.tool_state.mouse_state = mouse_state;
-
-				state.document.svg.push(SvgElement::Circle(Circle {
-					center: Point {
-						x: mouse_state.position.x as f64,
-						y: mouse_state.position.y as f64,
-					},
-					radius: 10.0,
-				}));
-				self.emit_response(Response::UpdateCanvas { document: state.document.render() });
-
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state = *mouse_state;
 			}
 			Event::MouseMovement(pos) => {
-				state.tool_state.mouse_state.position = pos;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state.position = *pos;
 			}
 			Event::ModifierKeyDown(mod_keys) => {
-				state.tool_state.mod_keys = mod_keys;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mod_keys = *mod_keys;
 			}
 			Event::ModifierKeyUp(mod_keys) => {
-				state.tool_state.mod_keys = mod_keys;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mod_keys = *mod_keys;
 			}
 			Event::KeyPress(key) => todo!(),
 		}
+
+		let (responses, operations) = tool_state.active_tool()?.handle_input(event, document);
+
+		self.dispatch_operations(document, &operations);
+		// TODO - Dispatch Responses
+
+		Ok(())
+	}
+
+	fn dispatch_operations(&self, document: &mut Document, operations: &[Operation]) {
+		for operation in operations {
+			self.dispatch_operation(document, operation);
+		}
+	}
+
+	fn dispatch_operation(&self, document: &mut Document, operation: &Operation) {
+		document.handle_operation(operation, |svg: String| {
+			self.dispatch_response(&Response::UpdateCanvas { document: svg });
+		});
+	}
+
+	pub fn dispatch_responses(&self, responses: &[Response]) {
+		for response in responses {
+			self.dispatch_response(response);
+		}
 	}
 
-	pub fn emit_response(&self, response: Response) {
+	pub fn dispatch_response(&self, response: &Response) {
 		let func = &self.callback;
-		func(response)
+		// TODO - Remove clone if possible
+		func(response.clone())
```

Comment:
This is bad API design. Instead, you want to keep the function as taking a `Response` and let the caller of this function clone that if needed.